### DISCUSSION TITLE
useFieldArray append not working properly

### DISCUSSION THREAD

**Discussion by ronaldruzicka** [react-hook-form/react-hook-form#2840] [Q&A] [answered]
Hi there, 

I'm having a problem with `append` when using `useFieldArray`. I need to add a new project from a select of options and after clicking on `CONFIRM` it should append the project to the form values. 

What it does is, that it logs the new project properly before `append`, but when I watch for "projects" it then logs `undefined` and then `[{ roles: undefined }]`, but not the actual values that were passed to `append`.

Am I doing something wrong? 
Please help 🙏🏻

Here is codesandbox: https://codesandbox.io/s/beautiful-banzai-b4jbk?file=/src/App.tsx

Thanks!

**Comment by bluebill1049** [comment, marked as answer]
each useFildArray host its own FieldArray, you will have to do an append at your child field array's.
```ts
useEffect(() => {
  append({})
}, [append])
```

more on this, next update you can use `setValue`: https://github.com/react-hook-form/react-hook-form/pull/2691

`setValue('fieldArray', whateverDataHere)`

**Reply by ronaldruzicka** [reply]
Thanks Bill!

But I have just one field array which is **projects**, I'm not using nested field arrays.
Or am I missing something?

**Reply by bluebill1049** [reply]
```
append({
        active: newProject.active,
        id: newProject.id,
        name: newProject.name,
        roles: [] // this was what i thought you have array of array field
      });
```

**Reply by ronaldruzicka** [reply]
Oh that, well the `roles` will be selected from the actual project as a multiselect from the `Autocomplete` component from MUI. There I'm not using field array. The problem is when you check the `selectedProjects` it should have `newProject` in there but it's not.

This is `newProject` which goes to `append`:
```js
{
  id: "2f10f400-832f-48ba-8141-5de2317b9cc4",
  dateCreated: "2020-05-11T06:49:44.261Z",
  dateUpdated: "2020-05-11T06:49:44.261Z",
  name: "first project",
  active: true,
  tenantId: "6eb0adf1-a409-40dc-b392-4b9dcd551b78",
}
```

And `selectedProjects` is logged as `undefined` and then:
```js
{
  roles: undefined
}
```

**Reply by ronaldruzicka** [reply]
I added a button that adds hardcoded project and it still won't work 😕

**Reply by bluebill1049** [reply]
so the issue is `watch` not return the correct field value?

**Reply by bluebill1049** [reply]
why are you having those inner components as well?

handleRenderProjectSelect
renderProjects

they get mount and remounted each render.

**Reply by ronaldruzicka** [reply]
Well I don't know if it's `append` or `watch`. Probably `append` because when I submit the form, the values are also wrong. So `watch` is just returning form values which are wrong.

**Reply by ronaldruzicka** [reply]
You mean why in a separate function?

**Reply by bluebill1049** [reply]
1. those inner components should be separate components instead live inside your main component (but that's not the issue)
2. here is working CSB: https://codesandbox.io/s/cocky-chaum-m73dl?file=/src/App.js maybe take a look at the diff.

**Reply by ronaldruzicka** [reply]
Thanks I'll check 👍

**Reply by ronaldruzicka** [reply]
Damn, I was missing `defaultValue`! 🤦‍♂️

Thanks for the help!

**Reply by bluebill1049** [reply]
np, but do follow my suggestion above, otherwise the perf is terrible for your form. your are unmount and remount all your child components inside.

**Reply by ronaldruzicka** [reply]
Yes, I'm refactoring right now and putting it in separate components.

**Comment by netrolite** [comment]
in my case i created a custom hook that was returning the return value of `useFieldArray` and i called that hook from multiple components. But the problem was that for it to work, every component had to be using the same object returned by the `useFieldArray` hook. Instead, every component received a different reference to the return value, I guess?

Basically I had to delete the hook entirely and pass the return value of `useFieldArray` from the parent component to all its children that needed it. Though I might want to instead use a context to avoid prop-drilling
